glean-core/src/upload/mod.rs

```diff
@@ -69,6 +77,74 @@ pub fn setup_upload_manager(upload_manager: PingUploadManager) -> Result<()> {
     Ok(())
 }
 
+/// A counter to help limit the number of processing pings per minute.
+///
+/// NOTE: A "processing" ping is a ping that has been dequeued through `get_upload_task`.
+#[derive(Debug)]
+struct PingCounter {
+    started: Option<SystemTime>,
+    count: u32,
+}
+
+impl PingCounter {
+    pub fn new() -> Self {
+        Self {
+            started: None,
+            count: 0,
+        }
+    }
+
+    fn reset(&mut self) {
+        self.started = Some(SystemTime::now());
+        self.count = 0;
+    }
+
+    /// The counter should reset if
+    ///
+    /// 1. It has never started;
+    /// 2. It has been started more than one minute ago;
+    /// 3. Something goes wrong while trying to calculate the elapsed time since the last reset.
+    fn should_reset(&self) -> bool {
+        if self.started.is_none() {
+            return true;
+        }
+
+        match self.started.unwrap().elapsed() {
+            Ok(elapsed) => {
+                let remaining = max(0, SECONDS_IN_A_MINUTE - elapsed.as_secs() as u32);
+                if remaining == 0 {
+                    return true;
+                }
+            }
+            Err(e) => {
+                log::warn!(
+                    "Error getting elapse time since start of ping counter: {:?}",
+                    e
+                );
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Increments the counter by 1 and returns the current count.
+    ///
+    /// This will return `None` if the counter has reached the maximum.
+    pub fn increment(&mut self) -> Option<u32> {
```

Comment:
No need to return the count, in hindsight I don't even know why I put that there. I didn't use a `Result` here just because I wanted to avoid creating a new error type for this. The enum approach is nicer, I'll do it and we can bikeshed on the names after I send updates.